src/librustc_macros/src/session_diagnostic.rs

```diff
@@ -0,0 +1,660 @@
+            self.kind = Some((kind, span));
+            Ok(())
+        } else {
+            let kind_str = |kind: &DiagnosticId| match kind {
+                DiagnosticId::Lint(..) => "lint",
+                DiagnosticId::Error(..) => "error",
+            };
+
+            let existing_kind = kind_str(&self.kind.as_ref().unwrap().0);
+            let this_kind = kind_str(&kind);
+
+            let msg = if this_kind == existing_kind {
+                format!("`{}` specified multiple times", existing_kind)
+            } else {
+                format!("`{}` specified when `{}` was already specified", this_kind, existing_kind)
+            };
+            throw_span_err!(span.unwrap(), &msg);
+        }
+    }
+
+    fn generate_field_code(
+        &mut self,
+        attr: &syn::Attribute,
+        info: FieldInfo<'_>,
+    ) -> Result<proc_macro2::TokenStream, SessionDiagnosticDeriveError> {
+        let field_binding = &info.binding.binding;
+
+        let option_ty = option_inner_ty(&info.ty);
+
+        let generated_code = self.generate_non_option_field_code(
+            attr,
+            FieldInfo {
+                vis: info.vis,
+                binding: info.binding,
+                ty: option_ty.unwrap_or(&info.ty),
+                span: info.span,
+            },
+        )?;
+        Ok(if option_ty.is_none() {
+            quote! { #generated_code }
+        } else {
+            quote! {
+                if let Some(#field_binding) = #field_binding {
+                    #generated_code
+                }
+            }
+        })
+    }
+
+    fn generate_non_option_field_code(
+        &mut self,
+        attr: &syn::Attribute,
+        info: FieldInfo<'_>,
+    ) -> Result<proc_macro2::TokenStream, SessionDiagnosticDeriveError> {
+        let diag = &self.diag;
+        let field_binding = &info.binding.binding;
+        let name = attr.path.segments.last().unwrap().ident.to_string();
+        let name = name.as_str();
+        // At this point, we need to dispatch based on the attribute key + the
+        // type.
+        let meta = attr.parse_meta()?;
+        Ok(match meta {
+            syn::Meta::NameValue(syn::MetaNameValue { lit: syn::Lit::Str(s), .. }) => {
+                let formatted_str = self.build_format(&s.value(), attr.span());
+                match name {
+                    "message" => {
+                        if type_matches_path(&info.ty, &["rustc_span", "Span"]) {
+                            quote! {
+                                #diag.set_span(*#field_binding);
+                                #diag.set_primary_message(#formatted_str);
+                            }
+                        } else {
+                            throw_span_err!(
+                                attr.span().unwrap(),
+                                "the `#[message = \"...\"]` attribute can only be applied to fields of type Span"
+                            );
+                        }
```

Comment:
Basically I believe that 

```suggestion
                        quote! {
                            #diag.set_span(*#field_binding);
                            #diag.set_primary_message(#formatted_str);
                        }
```

will work and give good error messages. I think there's no need for you to do any type checks

Reply:
The error message without intervention looks like this:
```
error[E0277]: the trait bound `rustc_span::MultiSpan: std::convert::From<std::string::String>` is not satisfied
  --> $DIR/session-derive-errors.rs:92:10
   |
LL | #[derive(SessionDiagnostic)]
   |          ^^^^^^^^^^^^^^^^^ the trait `std::convert::From<std::string::String>` is not implemented for `rustc_span::MultiSpan`
   |
   = help: the following implementations were found:
             <rustc_span::MultiSpan as std::convert::From<rustc_span::Span>>
             <rustc_span::MultiSpan as std::convert::From<std::vec::Vec<rustc_span::Span>>>
   = note: required because of the requirements on the impl of `std::convert::Into<rustc_span::MultiSpan>` for `std::string::String`
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```
The message itself is fine, but the span it's on isn't particularly helpful, unfortunately.  I wonder if there's some way I can just correct the span..

Reply:
Actually, I realised I can use `quote_span!` instead to get the span in the right place:
```
error[E0277]: the trait bound `rustc_span::MultiSpan: std::convert::From<std::string::String>` is not satisfied
  --> $DIR/session-derive-errors.rs:95:5
   |
LL | /     #[message = "this message is applied to a String field"]
LL | |
LL | |     name: String,
   | |________________^ the trait `std::convert::From<std::string::String>` is not implemented for `rustc_span::MultiSpan`
   |
   = help: the following implementations were found:
             <rustc_span::MultiSpan as std::convert::From<rustc_span::Span>>
             <rustc_span::MultiSpan as std::convert::From<std::vec::Vec<rustc_span::Span>>>
   = note: required because of the requirements on the impl of `std::convert::Into<rustc_span::MultiSpan>` for `std::string::String`
```

Reply:
Hmm, though using `quote_span` worked fine for `message`, it doesn't work for `label`:

https://github.com/jumbatm/rust/blob/session-diagnostic-derive-ideal-spans/src/test/ui-fulldeps/session-derive-errors.stderr#L122-L146

I believe this is because E0308 is trying to be a bit cleverer with the span it emits on, but ends up pointing to a Span that's within the generated code, so ends up just pointing to the `#[derive(SessionDiagnostic)]`.

Reply:
wait I'm confused. We're talking about `#field_binding` here, right? How does it differ for `label` and `message`? This is just about the spans pointing to the error struct, not the spans of the error itself.

Reply:
Yeah, this is about the span of the error message when, say, the `label` attribute is placed on a non-Span field. For this scenario, we're relying on the type error in the code generated by the macro to bubble up and have the span that the generated tokens were set to.

> How does it differ for label and message?

The generated code expects a `Span` field for `label`, but expects an `impl MultiSpan` field for `message`, so having the wrong-typed field for these two causes slightly different errors (an E0308 for if a `label` field is the wrong type, and E0277 if a `message` field is the wrong type). E0277 seems to happily point at whatever span I point it at using `quote_spanned!`, but E0308  ends up pointing at the top-level `#[derive(SessionDiagnostic)]` tokens (as shown in the link to the stderr above), even if I `quote_spanned!` with the same span, as below:

https://github.com/jumbatm/rust/blob/5a1969561c8f50bb15a16db17a7bad4c9a813e36/src/librustc_macros/src/session_diagnostic.rs#L390-L400

Reply:
aaah, thanks, I had a look, and the span used for the diagnostic comes from somewhere deep in the obligation logic. Let's take the diagnostics hit for now and figure out this type mismatch span problem independently.